mm2src/lp_ordermatch.rs

```diff
@@ -457,6 +464,71 @@ struct GetOrderbookRes {
     pubkey_orders: HashMap<String, GetOrderbookPubkeyItem>,
 }
 
+#[derive(Debug, Deserialize, Serialize)]
+struct BestOrdersRes {
+    orders: HashMap<String, Vec<OrderbookItemWithProof>>,
+}
+
+async fn process_best_orders_p2p_request(
+    ctx: MmArc,
+    coin: String,
+    action: BestOrdersAction,
+    volume: BigRational,
+) -> Result<Option<Vec<u8>>, String> {
+    let ordermatch_ctx = OrdermatchContext::from_ctx(&ctx).expect("ordermatch_ctx must exist at this point");
+    let orderbook = ordermatch_ctx.orderbook.lock().await;
+    let search_pairs_in = match action {
+        BestOrdersAction::Buy => &orderbook.pairs_existing_for_base,
+        BestOrdersAction::Sell => &orderbook.pairs_existing_for_rel,
+    };
+    let tickers = match search_pairs_in.get(&coin) {
+        Some(tickers) => tickers,
+        None => return Ok(None),
+    };
+    let mut result = HashMap::new();
+    let pairs: Vec<_> = tickers.iter().map(|ticker| (coin.clone(), ticker.clone())).collect();
+    for pair in pairs.iter() {
+        let orders = match orderbook.ordered.get(pair) {
+            Some(orders) => orders,
+            None => {
+                log::warn!("No orders for pair {:?}", pair);
+                continue;
+            },
+        };
+        let mut best_orders = vec![];
+        let mut collected_volume = BigRational::zero();
+        for ordered in orders {
+            match orderbook.order_set.get(&ordered.uuid) {
+                Some(o) => {
+                    match orderbook.orderbook_item_with_proof(o.clone()) {
+                        Ok(order_w_proof) => best_orders.push(order_w_proof),
+                        Err(e) => {
+                            log::error!("Error {:?} on proof generation for order {:?}", e, o);
+                            continue;
+                        },
+                    };
+                    let coin_volume = match action {
+                        BestOrdersAction::Buy => o.max_volume.clone(),
+                        BestOrdersAction::Sell => &o.max_volume * &o.price,
+                    };
+                    collected_volume += coin_volume;
+                    if collected_volume >= volume {
+                        break;
+                    }
+                },
+                None => {
+                    log::warn!("No order with uuid {:?}", ordered.uuid);
+                    continue;
+                },
+            };
+        }
+        result.insert(pair.1.clone(), best_orders);
```

Comment:
Considering the previous comment, we could replace `pair.1.clone()` with `pair.1`